doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+
+Operation: set all the Used cells to Reserving state (e<sub>2</sub> in cell state machine); overwrite the Reserved (e<sub>6</sub> in cell state machine) and Reserving cells (e<sub>3</sub> in cell state machine); set the Free cell to Reserved (e<sub>5</sub> in cell state machine).
+
+__e<sub>2</sub>__:
+
+Condition: all the cells allocated to the AG are Reserved.
+
+Operation: set all the cells to Used (e<sub>8</sub> in cell state machine).
+
+__e<sub>3</sub>__:
+
+Condition: all pods deleted.
+
+Operation: set all the cells allocated to the AG in Free state (e<sub>1</sub> in cell state machine).
+
+__e<sub>4</sub>__:
+
+Condition: all pods deleted.
+
+Operation: none.
+
+__e<sub>5</sub>__:
+
+Condition: a Reserving or Reserved cell in this AG is being overwritten (e<sub>3</sub> or e<sub>6</sub> in cell state machine).
+
+Operation: set all the other Reserving cells to Used (e<sub>4</sub> in cell state machine), Reserved cells to Free (e<sub>7</sub> in cell state machine).
+
+__e<sub>6</sub>__:
+
+Condition: a cell from Used to Reserving (e<sub>2</sub> in cell state machine)
+
+Operation: none.
+
+__e<sub>7</sub>__:
+
+Condition: AG deleted.
+
+Operation: set the Reserved cells to Free (e<sub>7</sub> in cell state machine), set the Reserving cells to Used (e<sub>4</sub> in cell state machine).
+
+## Cell State Machine
+
+Cell is the resource unit in HiveD. The figure below shows the state machine of cell. Note that here cells are _lowest-level physical cells_ (we record states only in these cells).
+
+![AG](cell-state-machine.png)
+
+### States
+
+__Free__: the cell is not allocated to any AG.
+
+__Used__: the cell is allocated to an AG, and the AG is running on this cell.
+
+__Reserved__: the cell is allocated to an AG, but there is no AG running on this cell (e.g., the AG is Preempting).
+
+__Reserving__: the cell is allocated to an AG, but another AG is still running on the cell (e.g., the first AG is Preempting the running AG).
+
+Note that the reservation of cells (Reserved and Reserving states) is not necessarily designed for preemptions (i.e., reserving resources for the Preempting AGs). In the future it is possible that we extend this mechanism to support other reservation mechanisms.
+
+Common cell life cycles:
+
+Free -> Used -> Free (no preemption involved)
+
+Free -> Used -> Reserving -> Reserved -> Used -> Free (preemption involved)
+
+### Events
+
+Note: “cell view” in the below operations represents the in-memory data structures for scheduling, e.g., free cell list, cell bindings, cell priorities. Allocating / releasing a cell in the cell view will modify the free cell list (and split / merge the cell), create / destroy cell bindings, and set / reset the cell priority. These changes are immediately visible to the subsequent scheduling cycles.
```

Comment:
Do we have to introduce the cell view concept in this doc?
allocate the cell to the AG in the cell view.
->
allocate the cell to the AG.

Is it also ok?

Reply:
We emphasize "allocate in cell view" is to differentiate from "AG is in Allocated state". Allocate in cell view does not mean the AG transitions to Allocated state.
I modified the note of cell view above to explain this